client_side/.infrastructure/infrastructure.sh

```diff
@@ -0,0 +1,358 @@
+
+  # ".noverify" is touched so that precmd does
+  # not attempt to verify user output on the "start_task" command that was
+  # written to `.command`.
+  touch "${INFRA_DIR}/.noverify"
+
+  # write_log does not need to be called because it is called by precmd.
+}
+
+# "Uninstalls" Bash Preexec by removing its triggers.
+# Remove all variable files created by the infrastructure
+# Stops the experiment completely by returning from the sourced scripts.
+end_experiment() {
+  # This effectively uninstalls Bash Pre-exec.
+  # Makes it so any commands typed after the experiment has ended will not be
+  # passed through preexec and precmd.
+  PROMPT_COMMAND=${PROMPT_COMMAND_OG}
+  trap - DEBUG
+
+  # Remove all variable files.
+  find ${INFRA_DIR} -type f -name ".*" -delete
+  cd "${EXP_DIR}"
+
+  echo "Congratulations! You have completed the interactive portion of the experiment."
+  echo "Please take some time to fill out the survey on the experiment home page."
+
+  return 0
+}
+
+# Resets the user's file system directory by syncing it with the
+# infrastructure's extracted file system directory.
+reset_fs() {
+  rsync --omit-dir-times --recursive --quiet --delete \
+    "${FS_SYNC_DIR}/" "${FS_DIR}"
+}
+
+# Prints out the treatment conditions for the experiment and optionally starts
+# training for the infrastructure and/or Tellina.
+#
+# If the experiment just started, infra_training will be started.
+# If the current treatment is "T", tellina_training will be started.
+#
+# Parameters:
+# $1: the half of the experiment to begin treatment for, can be 1 or 2.
+begin_treatment() {
+  # Sets the current treatment and task set based on the task ordering for the
+  # experiment.
+  local experiment_half=$1
+
+  if (( ${experiment_half} == 1 )); then
+    treatment="${TASK_ORDER:0:1}"
+    task_set=${TASK_ORDER:1:1}
+  else
+    treatment="${TASK_ORDER:2:1}"
+    task_set=${TASK_ORDER:3:1}
+  fi
+
+  print_treatment
+
+  if (( task_num == 0 )); then
+    if ! [[ -f "${INFRA_DIR}/.task_num" ]]; then
+      # If the user is at the very beginning and is not resuming to task 1, then
+      # enables infrastructure training.
+      INF_TRAINING=true
+
+      if [[ "$treatment" == "T" ]]; then
+        # If the treatment is T, enables Tellina training.
+        TEL_TRAINING=true
+      fi
+    fi
+  else
+    # If the treatment is T and the user is not at the beginning, enables
+    # Tellina training.
+    if [[ "$treatment" == "T" ]]; then
+      TEL_TRAINING=true
+    fi
+  fi
+}
+
+# Checks whether infrastructure training or Tellina training is enabled and
+# disables them if possible.
+#
+# Prioritizes infrastructure training.
+check_and_update_training_status() {
+  if [[ "${INF_TRAINING:-false}" == "true" ]]; then
+    if [[ "${status}" == "success" ]]; then
+      # If the user successfully did the infrastructure training, disables it.
+      unset INF_TRAINING
+
+      if [[ "${TEL_TRAINING:-false}" == "true" ]]; then
```

Comment:
The reason for the `elif` is so that infrastructure training does not happen at
the same time as Tellina training. If we change the `elif` into an `if`, the
next if will still have to check whether we are in `INF_TRAINING`. This does
reduce the duplication by a little bit, but adds another level for the nested `if`s.

Reply:
I think the problem here is that I don't fully understand how training works.  I think this code doesn't do the training, just sets things up so that later code will do the training.  Maybe a brief explanation of that would have headed off my confusion.

Reply:
I will add the explanation.